Fix recommendations for Scala versions and add better messages.

tgodzik opened the pull request:
Previously, we would recommened Scala 2.12.11 even if the user is already using 2.13. Also, when using an unsupported Scala version we would just say that project is misconfigured, which is the same in case of a missing semanticDB. 

Now, we recommend the last version for each binary except 2.11. And we also display if any Scala version is unsupported in the warning.

Closes https://github.com/scalameta/metals/issues/1548

![image](https://user-images.githubusercontent.com/3807253/77742488-d0a09e00-7016-11ea-9e09-597e1118f896.png)

ckipp01 commented on metals/src/main/scala/scala/meta/internal/metals/ScalaVersions.scala:
```diff
@@ -22,9 +22,18 @@ object ScalaVersions {
+        binaryVersion == scalaBinaryVersionFromFullVersion(latest)
+      )
+      .getOrElse {
+        BuildInfo.scala212
```
Just out of curiosity, if we do fall into this, why do we recommend 12 over 13?

tgodzik replied:
Good question, mostly because we are using it Metals I guess, so we are most sure about. We could do the newest available version.

ckipp01 replied:
I'm sort of on the side that if we are in this scenario, we should recommend 13. /shrug

tgodzik replied:
Switched to 2.13

tgodzik commented:
I added also a warning if a not yet supported version is released, which will be useful in any future upgrade. What do you all think?

ckipp01 commented:
> I added also a warning if a not yet supported version is released, which will be useful in any future upgrade. What do you all think?

Nice, that makes sense 👍

kpbochenek commented:
Awesome work! (LGTM!)

![2020-03-29-120044_1243x322_scrot](https://user-images.githubusercontent.com/10478402/77846255-109d8780-71b5-11ea-9173-aa1dd8e70870.png)

gabro commented:
One minor concern I have is about the 2.11 message: if you're on a minor version of 2.11 it might be bit strange to be told to upgrade to 2.13.

Nothing blocking anyway, truly great work!

tgodzik commented:
> One minor concern I have is about the 2.11 message: if you're on a minor version of 2.11 it might be bit strange to be told to upgrade to 2.13.
> 
> Nothing blocking anyway, truly great work!

I think you are right, let's recommend 2.11.12, first so that the user is less discouraged. A warning for later be displayed later on anyway.

tgodzik commented:
In case of 2.11.x messages  the message now looks:
```
You are using Scala versions 2.11.9, 2.12.1, 2.12.2, which are not supported in this version of Metals. Please upgrade to Scala versions 2.12.11, 2.13.1 or alternatively to legacy Scala 2.11.12.
```

I also refactored the logic for getting recommended versions and added a test suite, because I wanted to document what messages are being show in some particular cases. @gabro What do you think?

I am taking my time with this PR, because it's been very often that the messages were wrong here and users reported it often.

tgodzik commented:
When we refactor the Doctor we should most likely create an additional LSP suite for it to totally control the possible output.

ckipp01 commented on tests/unit/src/test/scala/tests/MessagesSuite.scala:
```diff
@@ -0,0 +1,47 @@
+      Messages.UnsupportedScalaVersion
+        .message(Seq("2.11.9", "2.12.2", "2.12.1")),
+      "You are using Scala versions 2.11.9, 2.12.1, 2.12.2, which are not supported in this version of Metals." +
+        " Please upgrade to Scala version 2.12.11 or alternatively to legacy Scala 2.11.12."
```
🤔 I guess I'm still not convinced that suggesting to a user that is currently on 2.12.2 that they switch to 2.11.12 is a good idea. In this specific situation, what would be the benefit of suggesting this current message over just `Please upgrade to Scala versions 2.12.11`?

tgodzik replied:
They are both on `2.11.9` and `2.12.2`, it might be another module they have in the project that is for Scala native and we cannot recommend they upgrade to 2.12. This should be a rare situation though.

ckipp01 replied:
ahhhh, sorry, I misunderstood the test. Yea, then this is fine.
